File: jscomp/test/js_re_test.ml

let suites = Mt.[
  "captures", (fun _ ->
    let re = [%re "/(\\d+)-(?:(\\d+))?/g"] in
    let str = "3-" in
    match re |. Js.Re2.exec str with
      | Some result ->
        let defined = (Js.Re2.captures result).(1) in
        let undefined = (Js.Re2.captures result).(2) in
        Eq((Js.Nullable.return "3", Js.Nullable.null), (defined, undefined))
      | None -> Fail()
  );

  "fromString", (fun _ ->
    (* From the example in js_re.mli *)
    let contentOf tag xmlString =
      Js.Re2.fromString ("<" ^ tag ^ ">(.*?)<\\/" ^ tag ^">")
        |. Js.Re2.exec xmlString
        |> function
          | Some result -> Js.Nullable.toOption (Js.Re2.captures result).(1)
          | None -> None in
    Eq (contentOf "div" "<div>Hi</div>", Some "Hi")
  );

  "exec_literal", (fun _ ->
    match [%re "/[^.]+/"] |. Js.Re2.exec "http://xxx.domain.com" with
    | Some res ->
      Eq(Js.Nullable.return "http://xxx", (Js.Re2.captures res).(0))
    | None ->
      FailWith "regex should match"
  );

  "exec_no_match", (fun _ ->
    match [%re "/https:\\/\\/(.*)/"] |. Js.Re2.exec "http://xxx.domain.com" with
    | Some _ ->  FailWith "regex should not match"
    | None -> Ok true
  );

  "test_str", (fun _ ->
    let res = "foo"
      |. Js.Re2.fromString
      |. Js.Re2.test "#foo#" in

    Eq(true, res)
  );

  "fromStringWithFlags", (fun _ ->
    let res = Js.Re2.fromStringWithFlags "foo" "g" in

    Eq(true, res |. Js.Re2.global)
  );
  "result_index", (fun _ ->
    match "zbar" |. Js.Re2.fromString |. Js.Re2.exec "foobarbazbar" with
    | Some res ->
      Eq(8, res |. Js.Re2.index)
    | None ->
      Fail ()
  );
  "result_input", (fun _ ->
    let input = "foobar" in

    match [%re "/foo/g"] |. Js.Re2.exec input with
    | Some res ->
      Eq(input,  res |. Js.Re2.input)
    | None ->
      Fail ()
  );

  (* es2015 *)
  "t_flags", (fun _ ->
    Eq("gi", [%re "/./ig"] |. Js.Re2.flags)
  );

  "t_global", (fun _ ->
    Eq(true, [%re "/./ig"] |. Js.Re2.global)
  );
  "t_ignoreCase", (fun _ ->
    Eq(true, [%re "/./ig"] |. Js.Re2.ignoreCase)
  );
  "t_lastIndex", (fun _ ->
    let re = [%re "/na/g"] in
    let _ = re |. Js.Re2.exec "banana" in

    Eq(4,  re |. Js.Re2.lastIndex)
  );
  "t_setLastIndex", (fun _ ->
    let re = [%re "/na/g"] in

    let before = Js.Re2.lastIndex re in
    let () = Js.Re2.setLastIndex re 42 in
    let after = Js.Re2.lastIndex re in

    Eq((0, 42),  (before, after))
  );
  "t_multiline", (fun _ ->
    Eq(false, [%re "/./ig"] |. Js.Re2.multiline)
  );
  "t_source", (fun _ ->
    Eq("f.+o", [%re "/f.+o/ig"] |. Js.Re2.source)
  );

  (* es2015 *)
  "t_sticky", (fun _ ->
    Eq(true, [%re "/./yg"] |. Js.Re2.sticky)
  );
  "t_unicode", (fun _ ->
    Eq(false, [%re "/./yg"] |. Js.Re2.unicode)
  );
]

;; Mt.from_pair_suites __MODULE__ suites

File: jscomp/test/js_re_test.js
'use strict';

var Mt = require("./mt.js");
var Caml_array = require("../../lib/js/caml_array.js");
var Caml_option = require("../../lib/js/caml_option.js");

var suites_0 = [
  "captures",
  (function (param) {
      var re = /(\d+)-(?:(\d+))?/g;
      var result = re.exec("3-");
      if (result === null) {
        return {
                TAG: /* Fail */8,
                _0: undefined
              };
      }
      var defined = Caml_array.get(result, 1);
      var $$undefined = Caml_array.get(result, 2);
      return {
              TAG: /* Eq */0,
              _0: [
                "3",
                null
              ],
              _1: [
                defined,
                $$undefined
              ]
            };
    })
];

var suites_1 = {
  hd: [
    "fromString",
    (function (param) {
        var contentOf = function (tag, xmlString) {
          var result = new RegExp("<" + (tag + (">(.*?)<\\/" + (tag + ">")))).exec(xmlString);
          if (result !== null) {
            return Caml_option.nullable_to_opt(Caml_array.get(result, 1));
          }
          
        };
        return {
                TAG: /* Eq */0,
                _0: contentOf("div", "<div>Hi</div>"),
                _1: "Hi"
              };
      })
  ],
  tl: {
    hd: [
      "exec_literal",
      (function (param) {
          var res = /[^.]+/.exec("http://xxx.domain.com");
          if (res !== null) {
            return {
                    TAG: /* Eq */0,
                    _0: "http://xxx",
                    _1: Caml_array.get(res, 0)
                  };
          } else {
            return {
                    TAG: /* FailWith */9,
                    _0: "regex should match"
                  };
          }
        })
    ],
    tl: {
      hd: [
        "exec_no_match",
        (function (param) {
            var match = /https:\/\/(.*)/.exec("http://xxx.domain.com");
            if (match !== null) {
              return {
                      TAG: /* FailWith */9,
                      _0: "regex should not match"
                    };
            } else {
              return {
                      TAG: /* Ok */4,
                      _0: true
                    };
            }
          })
      ],
      tl: {
        hd: [
          "test_str",
          (function (param) {
              var res = new RegExp("foo").test("#foo#");
              return {
                      TAG: /* Eq */0,
                      _0: true,
                      _1: res
                    };
            })
        ],
        tl: {
          hd: [
            "fromStringWithFlags",
            (function (param) {
                var res = new RegExp("foo", "g");
                return {
                        TAG: /* Eq */0,
                        _0: true,
                        _1: res.global
                      };
              })
          ],
          tl: {
            hd: [
              "result_index",
              (function (param) {
                  var res = new RegExp("zbar").exec("foobarbazbar");
                  if (res !== null) {
                    return {
                            TAG: /* Eq */0,
                            _0: 8,
                            _1: res.index
                          };
                  } else {
                    return {
                            TAG: /* Fail */8,
                            _0: undefined
                          };
                  }
                })
            ],
            tl: {
              hd: [
                "result_input",
                (function (param) {
                    var input = "foobar";
                    var res = /foo/g.exec(input);
                    if (res !== null) {
                      return {
                              TAG: /* Eq */0,
                              _0: input,
                              _1: res.input
                            };
                    } else {
                      return {
                              TAG: /* Fail */8,
                              _0: undefined
                            };
                    }
                  })
              ],
              tl: {
                hd: [
                  "t_flags",
                  (function (param) {
                      return {
                              TAG: /* Eq */0,
                              _0: "gi",
                              _1: /./ig.flags
                            };
                    })
                ],
                tl: {
                  hd: [
                    "t_global",
                    (function (param) {
                        return {
                                TAG: /* Eq */0,
                                _0: true,
                                _1: /./ig.global
                              };
                      })
                  ],
                  tl: {
                    hd: [
                      "t_ignoreCase",
                      (function (param) {
                          return {
                                  TAG: /* Eq */0,
                                  _0: true,
                                  _1: /./ig.ignoreCase
                                };
                        })
                    ],
                    tl: {
                      hd: [
                        "t_lastIndex",
                        (function (param) {
                            var re = /na/g;
                            re.exec("banana");
                            return {
                                    TAG: /* Eq */0,
                                    _0: 4,
                                    _1: re.lastIndex
                                  };
                          })
                      ],
                      tl: {
                        hd: [
                          "t_setLastIndex",
                          (function (param) {
                              var re = /na/g;
                              var before = re.lastIndex;
                              re.lastIndex = 42;
                              var after = re.lastIndex;
                              return {
                                      TAG: /* Eq */0,
                                      _0: [
                                        0,
                                        42
                                      ],
                                      _1: [
                                        before,
                                        after
                                      ]
                                    };
                            })
                        ],
                        tl: {
                          hd: [
                            "t_multiline",
                            (function (param) {
                                return {
                                        TAG: /* Eq */0,
                                        _0: false,
                                        _1: /./ig.multiline
                                      };
                              })
                          ],
                          tl: {
                            hd: [
                              "t_source",
                              (function (param) {
                                  return {
                                          TAG: /* Eq */0,
                                          _0: "f.+o",
                                          _1: /f.+o/ig.source
                                        };
                                })
                            ],
                            tl: {
                              hd: [
                                "t_sticky",
                                (function (param) {
                                    return {
                                            TAG: /* Eq */0,
                                            _0: true,
                                            _1: /./yg.sticky
                                          };
                                  })
                              ],
                              tl: {
                                hd: [
                                  "t_unicode",
                                  (function (param) {
                                      return {
                                              TAG: /* Eq */0,
                                              _0: false,
                                              _1: /./yg.unicode
                                            };
                                    })
                                ],
                                tl: /* [] */0
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
};

var suites = {
  hd: suites_0,
  tl: suites_1
};

Mt.from_pair_suites("Js_re_test", suites);

exports.suites = suites;
/*  Not a pure module */
